api/example.com/resource/gpu/v1alpha1/sharing.go

```diff
@@ -0,0 +1,121 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	"fmt"
+)
+
+// These constants represent the different Sharing strategies.
+const (
+	TimeSlicingStrategy       GpuSharingStrategy = "TimeSlicing"
+	SpacePartitioningStrategy GpuSharingStrategy = "SpacePartitioning"
+)
+
+// These constants represent the different TimeSlicing configurations.
+const (
+	DefaultTimeSlice TimeSliceDuration = "Default"
+	ShortTimeSlice   TimeSliceDuration = "Short"
+	MediumTimeSlice  TimeSliceDuration = "Medium"
+	LongTimeSlice    TimeSliceDuration = "Long"
+)
+
+// Sharing provides methods to check if a given sharing strategy is selected
+// and grab its configuration.
+// +k8s:deepcopy-gen=false
+type Interface interface {
+	IsTimeSlicing() bool
+	IsSpacePartitioning() bool
+	GetTimeSlicingConfig() (*TimeSlicingConfig, error)
+	GetSpacePartitioningConfig() (*SpacePartitioningConfig, error)
+}
+
+// GpuSharingStrategy encodes the valid Sharing strategies as a string.
+// +kubebuilder:validation:Enum=TimeSlicing;SpacePartitioning
+type GpuSharingStrategy string
+
+// TimeSliceDuration encodes the valid timeslice duration as a string.
+// +kubebuilder:validation:Enum=Default;Short;Medium;Long
+type TimeSliceDuration string
+
+// GpuSharing holds the current sharing strategy for GPUs and its settings.
+// +kubebuilder:validation:MaxProperties=2
+type GpuSharing struct {
+	// +kubebuilder:default=TimeSlicing
+	// +kubebuilder:validation:Required
+	Strategy                GpuSharingStrategy       `json:"strategy"`
+	TimeSlicingConfig       *TimeSlicingConfig       `json:"timeSlicingConfig,omitempty"`
+	SpacePartitioningConfig *SpacePartitioningConfig `json:"spacePartitioningConfig,omitempty"`
+}
+
+// TimeSlicingSettings provides the settings for CUDA time-slicing.
+type TimeSlicingConfig struct {
+	// +kubebuilder:default=Default
+	Duration TimeSliceDuration `json:"duration,omitempty"`
+}
+
+// SpacePartitioningConfig provides the configuring for the SpacePartitioning strategy.
+type SpacePartitioningConfig struct {
+	// SliceCount indicates how many equally sized (memory and compute) slices
+	// the GPU should be divided into. Each client that attaches will get
+	// access to exactly one of these slices.
+	PartitionCount int `json:"partitionCount,omitempty"`
+}
+
+// IsTimeSlicing checks if the TimeSlicing strategy is applied.
+func (s *GpuSharing) IsTimeSlicing() bool {
+	if s == nil {
+		return false
```

Comment:
I thought we said that timeslicing was the default. This seems to indicate that this is not the case?

Reply:
I decided to change it for the purpose of this function. The default gets set elsewhere explicitly (such that this would error if it wasn't for some reason).

Reply:
https://github.com/kubernetes-sigs/dra-example-driver/pull/54/files#diff-e8cc6726d33b90acdb66a8ebbc67662470c2f60da033553fc14d5dca412a98a5R181